charms/istio-pilot/src/istioctl.py

```diff
@@ -0,0 +1,214 @@
+        """
+        self._istioctl_path = istioctl_path
+        self._namespace = namespace
+        self._profile = profile
+
+    @property
+    def _istioctl_flags(self):
+        return [
+            "-s",
+            f"profile={self._profile}",
+            "-s",
+            f"values.global.istioNamespace={self._namespace}",
+        ]
+
+    def install(self):
+        """Wrapper for the `istioctl install` command."""
+        try:
+            subprocess.check_call([self._istioctl_path, "install", "-y", *self._istioctl_flags])
+        except subprocess.CalledProcessError as cpe:
+            error_msg = f"Failed to install istio using istioctl.  Exit code: {cpe.returncode}."
+            logging.error(error_msg)
+            logging.error(f"stdout: {cpe.stdout}")
+            logging.error(f"stderr: {cpe.stderr}")
+
+            raise InstallFailedError(error_msg) from cpe
+
+    def manifest(self) -> str:
+        """Wrapper for the `istioctl manifest generate` command.
+
+        Returns:
+            (str) a YAML string of the Kubernetes manifest for Istio
+        """
+        try:
+            manifests = subprocess.check_output(
+                [self._istioctl_path, "manifest", "generate", *self._istioctl_flags]
+            )
+        except subprocess.CalledProcessError as cpe:
+            error_msg = (
+                f"Failed to generate manifests for istio using istioctl. "
+                f"Exit code: {cpe.returncode}."
+            )
+            logging.error(error_msg)
+            logging.error(f"stdout: {cpe.stdout}")
+            logging.error(f"stderr: {cpe.stderr}")
+
+            raise ManifestFailedError(error_msg) from cpe
+
+        return manifests
+
+    def precheck(self):
+        """Executes `istioctl x precheck` to validate whether the environment can be updated.
+
+        Raises:
+            PrecheckFailedError: if the precheck command fails.
+        """
+        try:
+            subprocess.check_call(
+                [
+                    self._istioctl_path,
+                    "x",
+                    "precheck",
+                ]
+            )
+        except subprocess.CalledProcessError as cpe:
+            raise PrecheckFailedError(
+                "Upgrade failed during `istio precheck` with error code" f" {cpe.returncode}"
+            ) from cpe
+
+    def remove(self):
+        """Removes the Istio installation using istioctl and Lightkube.
+
+        TODO: Should we use `istioctl x uninstall` here instead of lightkube?  It is an
+        experimental feature but included in all istioctl versions we support.
+        """
+        manifest = self.manifest()
+
+        # Render YAML into Lightkube Objects
+        k8s_objects = codecs.load_all_yaml(manifest, create_resources_for_crds=True)
+
+        client = Client()
+        delete_many(client=client, objs=k8s_objects)
+
+    def upgrade(self, precheck: bool = True):
```

Comment:
Would upgrading using this method require us to update the istioctl binary that we set in charmcraft.yaml? 
Also, this seems to me like we don't really need any other charm to upgrade to. How does this method integrate with `juju refresh istio-pilot --channel <channel>`?

Reply:
The `Istioctl` class only contains how to operate **an** `istioctl`, nothing specific about a particular version of istioctl.  Upgrading using this method uses whatever istioctl binary you've given it to try to upgrade the k8s cluster.  

For the istio-pilot 1.16.2 charm, we'd include istioctl 1.16.2 in charmcraft.yaml (and similarly for 1.11.0, we'd include istioctl 1.11.0).  For this reason, yes atm we do still need a charm for each istio-pilot version.  We *could* add multiple binaries into the same charm (so istio-pilot 1.16.2 could also have older binaries included), but then we'd get much larger images because of the extra binaries, etc

Reply:
I think we've sorted this out offline and this can be resolved, but please reopen if I'm incorrect!